ignite/engines/engine.py

```diff
@@ -57,12 +59,40 @@ def add_event_handler(self, event_name, handler, *args, **kwargs):
             self._logger.error("attempt to add event handler to an invalid event %s ", event_name)
             raise ValueError("Event {} is not a valid event for this Engine".format(event_name))
 
+        self._check_handler_signature(handler, *args, **kwargs)
+
         if event_name not in self._event_handlers:
             self._event_handlers[event_name] = []
 
         self._event_handlers[event_name].append((handler, args, kwargs))
         self._logger.debug("added handler for event %s ", event_name)
 
+    def _check_handler_signature(self, handler, *args, **kwargs):
+        exception_msg = None
+
+        if sys.version_info[0] < 3:
```

Comment:
@sean-adler  what do you think about checking version globally and once instead of every time in `add_handler`, something like this: 
```python

if sys.version_info[0] < 3:
    
    def _check_handler_signature(self, handler, *args, **kwargs):
        exception_msg = None
        try:
            callable_ = handler if hasattr(handler, '__name__') else handler.__call__
            inspect.getcallargs(callable_, self, *args, **kwargs)
        except TypeError as exc:
            spec = inspect.getargspec(callable_)
            handler_params = list(spec.args)
            exception_msg = str(exc)
            
        if exception_msg:
            passed_params = [self] + list(args) + list(kwargs)
            raise ValueError("Error adding handler '{}': "
                             "takes parameters {} but will be called with {} "
                             "({})".format(
                                 handler, handler_params, passed_params, exception_msg))        
else:
    
    def _check_handler_signature(self, handler, *args, **kwargs):
        exception_msg = None
        signature = inspect.signature(handler)
        try:
            signature.bind(self, *args, **kwargs)
        except TypeError as exc:
            handler_params = list(signature.parameters)
            exception_msg = str(exc)

        if exception_msg:
            passed_params = [self] + list(args) + list(kwargs)
            raise ValueError("Error adding handler '{}': "
                             "takes parameters {} but will be called with {} "
                             "({})".format(
                                 handler, handler_params, passed_params, exception_msg))
```
but in a more compact way if possible :)

Reply:
The conditional is cheap to evaluate on every handler addition, I'm not sure it's worth complicating the code for?

Reply:
@alykhantejani I see your point, yes it is cheap to evaluate, but just esthetically I would tend to make a choice once...

A compacted version without `if` in `add_handler` as I see is like :
```python
# Define the following outside Engine
if sys.version_info[0] < 3:
    def _check_signature(handler, engine, *args, **kwargs):
        exception_msg = None
        callable_ = handler if hasattr(handler, '__name__') else handler.__call__
        try:
            inspect.getcallargs(callable_, engine, *args, **kwargs)
        except TypeError as exc:
            exception_msg = str(exc)
        return inspect.getargspec(callable_), exception_msg
else:
    def _check_signature(handler, engine, *args, **kwargs):
        exception_msg = None
        signature = inspect.signature(handler)
        try:
            signature.bind(engine, *args, **kwargs)
        except TypeError as exc:
            exception_msg = str(exc)
        return list(signature.parameters), exception_msg


# A method in Engine
def check_handler_signature(engine, handler, *args, **kwargs):
    handler_params, exception_msg = _check_signature(handler, engine, *args, **kwargs)
    if exception_msg:
        passed_params = [engine] + list(args) + list(kwargs)
        raise ValueError("Error adding handler '{}': "
                         "takes parameters {} but will be called with {} "
                         "({})".format(handler, handler_params, passed_params, exception_msg))
```
Any comments, feedbask ?

Reply:
both approaches are fine with me tbh :)

Reply:
Thanks @alykhantejani and @vfdev-5 - I took the middle-ground and moved this conditional into a constant 😄